Bug#99975 Request for help - Existing user is able to request for an organization/position

tarentomaheshvakkund opened the pull request:
1. Added a configuration- sbApiKey i.e the API  Key.
2. Added Constant values in the constant files.
3. Added a SunbirdApiRequest model class for the API Request.
4. Added isUserDetailExists in the OrganisationWorkFlowServiceImpl which is responsible for validating the user details if they exist in the system.
5. Added sb.api.key in the application properties.
6.bug #99966 -  Request for help - User is able to request for an existing Org/Position/Domain

tarentomaheshvakkund committed:
Bug#99975 Request for help - Existing user is able to request for an organization/position

1. Added a configuration- sbApiKey i.e the API  Key.
2. Added Constant values in the constant files.
3. Added a SunbirdApiRequest model class for the API Request.
4. Added isUserDetailExists in the OrganisationWorkFlowServiceImpl which is responsible for validating the user details if they exist in the system.
5. Added sb.api.key in the application properties.

karthik-tarento commented on src/main/java/org/sunbird/workflow/service/impl/OrganisationWorkFlowServiceImpl.java:
```diff
@@ -1,23 +1,65 @@
-        Response response = workflowService.workflowTransition(rootOrg, org, wfRequest);
+        Response response;
+        wfRequest.getUpdateFieldValues().forEach(updateFieldValuesMap -> updateFieldValuesMap.forEach((updateFieldValuesKey, updateFieldValuesValue) -> {
+            if (updateFieldValuesKey.equalsIgnoreCase(Constants.EMAIL) && isUserDetailExists(Constants.EMAIL, updateFieldValuesValue.toString())) {
```
you should always use the constant at first... the code suppose to be,
Constants.EMAIL.equalsIgnoreCase(updateFieldValuesKey)

tarentomaheshvakkund replied:
1.Constants are used first to check to avoid the NullPointerException.
2.Added all the string variables in the Constants.java file.

karthik-tarento commented on src/main/java/org/sunbird/workflow/service/impl/OrganisationWorkFlowServiceImpl.java:
```diff
@@ -1,23 +1,65 @@
+        Response response;
+        wfRequest.getUpdateFieldValues().forEach(updateFieldValuesMap -> updateFieldValuesMap.forEach((updateFieldValuesKey, updateFieldValuesValue) -> {
+            if (updateFieldValuesKey.equalsIgnoreCase(Constants.EMAIL) && isUserDetailExists(Constants.EMAIL, updateFieldValuesValue.toString())) {
+                throw new InvalidDataInputException(Constants.EMAIL_EXIST_ERROR);
```
Should not throw exception... it should return 400 bad request which is updated properly in response object.

tarentomaheshvakkund replied:
1. Removed the throw new Exception code block.
2. Made an entry in the response object with the status and the error message.
3. Based on the status if its a Bad request Response is returned accordingly else the workflow transition is called.

karthik-tarento commented on src/main/java/org/sunbird/workflow/service/impl/OrganisationWorkFlowServiceImpl.java:
```diff
@@ -1,23 +1,65 @@
+            } else if (updateFieldValuesKey.equalsIgnoreCase(Constants.TO_VALUE)) {
+                Map<String, Object> toValueMap = (Map<String, Object>) updateFieldValuesValue;
+                toValueMap.forEach((toValueKey, toValueValue) -> {
+                    if (toValueKey.equalsIgnoreCase(Constants.ORGANISATION_SERVICE_NAME) && isUserDetailExists(Constants.ORGANIZATION_NAME, toValueValue.toString())) {
```
Need to use different Api for org search.

tarentomaheshvakkund replied:
1. Created a new service method isOrgDetailExists and added the end point in the application.properties.
2. Removed the SunbirdApiRequest model class and created the request body using an additional HashMap.

tarentomaheshvakkund committed:
Bug#99975 - Code Review Comments fix.

1. Constants are used prior while comparing to avoid null pointer exception.
2. Removed the Throw new Exception object and added the errormessages and status in the response object.
3. Removed the SunbirdApiRequest Model class and handled the request creation by adding the request into  a map.
4. Added a new service method isOrgDetailExists to check for the organization details.
5. Added a new lms.org.search.path in the configuration file and the application properties for the orgsearch end point.

karthik-tarento commented on src/main/java/org/sunbird/workflow/service/impl/OrganisationWorkFlowServiceImpl.java:
```diff
@@ -38,4 +87,65 @@ public Response orgSearch(String rootOrg, String org, SearchCriteria criteria) {
+        requestObj.put(Constants.REQUEST, reqMap);
+        HashMap<String, String> headersValue = new HashMap<>();
+        headersValue.put(Constants.CONTENT_TYPE, Constants.APPLICATION_JSON);
+        headersValue.put(Constants.AUTHORIZATION, configuration.getSbApiKey());
```
check authorization is required

tarentomaheshvakkund replied:
Without Authorization the API is fetching the results so removed the Authorization from the header.

karthik-tarento commented on src/main/java/org/sunbird/workflow/service/impl/OrganisationWorkFlowServiceImpl.java:
```diff
@@ -38,4 +87,65 @@ public Response orgSearch(String rootOrg, String org, SearchCriteria criteria) {
+        requestObj.put(Constants.REQUEST, reqMap);
+        HashMap<String, String> headersValue = new HashMap<>();
+        headersValue.put(Constants.CONTENT_TYPE, Constants.APPLICATION_JSON);
+        headersValue.put(Constants.AUTHORIZATION, configuration.getSbApiKey());
```
check authorization is required

tarentomaheshvakkund replied:
1.Without Authorization the API is fetching the results so removed the Authorization from the header.

karthik-tarento commented on src/main/java/org/sunbird/workflow/service/impl/OrganisationWorkFlowServiceImpl.java:
```diff
@@ -1,24 +1,73 @@
-        return response;
+        Response response = new Response();
+        wfRequest.getUpdateFieldValues().forEach(updateFieldValuesMap -> updateFieldValuesMap.forEach((updateFieldValuesKey, updateFieldValuesValue) -> {
+            if (Constants.EMAIL.equalsIgnoreCase(updateFieldValuesKey) && isUserDetailExists(Constants.EMAIL, updateFieldValuesValue.toString())) {
```
updateFieldValuesValue.toString() --> .toString() method should not be called. This creates the chance for NPE.

(String) updateFieldValuesValue --> this is the correct approach.

tarentomaheshvakkund replied:
1. Casted the values to String and removed the toString() to avoid NPE.

tarentomaheshvakkund committed:
Code Review Fixes

1. Removed the sbApiKey from the configuration file and the application.properties.
2.Removed the AUTHORIZATION from the constants file.
3.Casted the valeus to the String from toString() to avoid nullPointerException.
4. Removed the Authorization in the headers- tested using portForwarding its working without API Key.
